/* [START declaration] */
syntax = "proto3";
package com.vmware.blockchain.deployment.model;

import "core.proto";
/* [END declaration] */

/* [START java_declaration] */
option java_package = "com.vmware.blockchain.deployment.model";
option java_multiple_files = true;
/* [END java_declaration] */

/* [START messages] */
/**
 * Denote the identifier of an orchestration site.
 *
 * @property[low]
 *   lower order bits of a 128-bit identifier;
 * @property[high]
 *   higher order bits of a 128-bit identifier;
 */
message OrchestrationSiteIdentifier {
    uint64 low = 1;
    uint64 high = 2;
}

/**
 * Model definition of information pertaining to an [OrchestrationSite].
 *
 * @param[type]
 *   type of the the endpoint.
 * @param[labels]
 *   a collection of key-value entries containing metadata information.
 * @param[vmc]
 *   information pertaining to a [Type.VMC] type endpoint.
 */
message OrchestrationSiteInfo {
    enum Type {
        NONE = 0;
        VMC = 1;
    }

    Type type = 1;
    map<string, string> labels = 2;
    VmcOrchestrationSiteInfo vmc = 3;
}

/**
 * Model definition of an orchestration site.
 *
 * @param[id]
 *   identifier of the site.
 * @param[info]
 *   information about the site.
 */
message OrchestrationSite {
    OrchestrationSiteIdentifier id = 1;
    OrchestrationSiteInfo info = 2;
}

/**
 * Model definition of an reportable view of an orchestration site.
 *
 * @param[id]
 *   identifier of the site.
 * @param[type]
 *   type of the the endpoint.
 * @param[labels]
 *   a collection of key-value entries containing metadata information.
 */
message OrchestrationSiteView {
    OrchestrationSiteIdentifier id = 1;
    OrchestrationSiteInfo.Type type = 2;
    map<string, string> labels = 3;
}

/**
 * Model definition of a VMware Cloud-based orchestration site.
 *
 * @param[authentication]
 *   authentication endpoint information.
 * @param[api]
 *   API endpoint information.
 * @param[containerRegistry]
 *   container registry endpoint information.
 * @param[organization]
 *   VMC organization identifier.
 * @param[datacenter]
 *   SDDC datacenter identifier owned by the organization.
 * @param[resourcePool]
 *   name of the resource pool to use compute deployment.
 * @param[folder]
 *   name of the inventory folder to use compute deployment.
 */
message VmcOrchestrationSiteInfo {
    Endpoint authentication = 1;
    Endpoint api = 2;
    Endpoint container_registry = 3;
    string organization = 4;
    string datacenter = 5;
    string resource_pool = 6;
    string folder = 7;
    string control_network = 8;
    fixed32 control_network_prefix = 9;
    int32 control_network_subnet = 10;
}
/* [END messages] */
